leaper.py

```diff
@@ -125,6 +123,26 @@ def _check_same_origin(self, origin, project):
         return project.startswith(origin)
 
     def check_source_submission(self, src_project, src_package, src_rev, target_project, target_package):
+        ret = self.check_source_submission_inner(src_project, src_package, src_rev, target_project, target_package)
+
+        # The layout of this code is just plain wrong and awkward. What is
+        # really desired a "post check source submission" not
+        # check_one_request() which applies to all action types and all actions
+        # at once. The follow-ups need the same context that the main method
+        # determining to flip the switch had. For maintenance incidents the
+        # ReviewBot code does a fair bit of mangling to the package and project
+        # values passed in which is not present during check_one_request().
+        # Currently the only feature used by maintenance is
+        # do_check_maintainer_review and the rest of the checks apply in the
+        # single action workflow so this should work fine, but really all the
+        # processing should be done per-action instead of per-request.
```

Comment:
I don't get your rant. If leaper has code in check_one_request() that needs to be moved to check_source_submission() or even to action specific functions then what prevents us from doing that? Also clearly the existing check_source_submission() in leaper is a mess and needs refactoring.

Reply:
It's mixing overall (ie all action) follow-ups and per-action toggle follow-ups. It could be moved, but for example it had to be moved for devel project review stuff since the loop on actions was in the removed code so it needed to be in the check_source_submission() since that is per-action.

This was always a problem, but since we enforce single-actions for the non-maintenance projects it is never exposed.

Reply:
To be clear, this means for a request with > 1 action it adds the devel reviews for each action an additional time per action. Meaning `add_review()` is called `len(actions) * len(actions)`. Obviously if `len(actions) = 1` it's not so much of an issue and since I made `add_review()` avoid duplicates it also would hide the negative side effects, but for maintenance where they commonly have >10 actions you would see 100 reviews added. :)

Reply:
As an side note, this whole process seems wrong since we do not really want to add "devel project" review, we want to add "expected origin" review. Which in many cases in devel, but not always. This was attempted to be handled via `/search/owner`, but that just falls back to priority list of OBS, in the case where the product origin differs from Factory the reviewer is wrong. This has complicated some of the weird results seen by leaper. I'll make an issue under Origin Manager milestone to consider this.